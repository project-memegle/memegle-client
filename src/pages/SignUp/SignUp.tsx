import { FormEvent, useCallback, useRef, useState } from 'react';
import validateId from 'components/Validations/ValidateId';
import validateNickname from 'components/Validations/ValidateNickname';
import { SignUpDTO } from 'services/dto/SignUpDto';
import { errorInputCheck } from 'utils/Event/errorInputCheck';
import handleInputChange from 'utils/Event/handleInputChange';
import passwordCheckHandler from 'utils/SignUp/passwordCheckHandler';
import useCustomNavigate from 'hooks/useCustomNaviaget';
import { signUp } from 'services/SignupService';
import { LogInRequestDTO } from 'services/dto/LogInDto';
import { logIn } from 'services/LogInService';
import { useAuth } from 'components/auth/ProvideAuth';
import { setSessionStorages } from 'utils/Storage/sessionStorage';
import StorageKeyword from 'Constant/StorageKeyword';
import getValidationMessages from 'components/Validations/ValidationMessages';
import { useTranslation } from 'react-i18next';
import { checkNickname } from 'services/NicknameService';
import { checkId } from 'services/IdService';
import { InputField } from 'components/UI/InputField';

interface ErrorMessageProps {
    message: string;
}

const ErrorMessage = ({ message }: ErrorMessageProps) =>
    message && <p className="error-message">{message}</p>;

interface SuccessMessageProps {
    message: string;
}

const SuccessMessage = ({ message }: SuccessMessageProps) =>
    message && <p className="success-message">{message}</p>;

interface ButtonProps {
    className: string;
    type: 'button' | 'submit';
    onClick?: (e: FormEvent<HTMLButtonElement>) => void;
    children: React.ReactNode;
}

const Button = ({ className, type, onClick, children }: ButtonProps) => (
    <button className={className} type={type} onClick={onClick}>
        {children}
    </button>
);

export default function SignUp() {
    const navigate = useCustomNavigate();
    const auth = useAuth();
    const { t } = useTranslation();
    const ValidationMessages = getValidationMessages();
    const DEFAULT_ID = ValidationMessages.DEFAULT_ID;
    const DEFAULT_NICKNAME = ValidationMessages.DEFAULT_NICKNAME;
    const DEFAULT_PASSWORD = ValidationMessages.DEFAULT_PASSWORD;
    // ---------------------ID----------------------------
    const idInputRef = useRef<HTMLInputElement>(null);
    const [id, setId] = useState<string>('');
    const [idErrorMessage, setIdErrorMessage] = useState<string>('');
    const [idSuccessMessage, setIdSuccessMessage] = useState<string>('');
    const [isIdDupliacted, setIsIdDupliacted] = useState<boolean>(false);
    const [isIdChecked, setIsIdChecked] = useState<boolean>(false);
    // ---------------------NICKNAME----------------------
    const nicknameInputRef = useRef<HTMLInputElement>(null);
    const [nickname, setNickname] = useState<string>('');
    const [nicknameErrorMessage, setNicknameErrorMessage] =
        useState<string>('');
    const [nicknameSuccessMessage, setNicknameSuccessMessage] =
        useState<string>('');
    const [isNicknameDupliacted, setIsNicknameDupliacted] =
        useState<boolean>(false);
    const [isNicknameChecked, setIsNicknameChecked] = useState<boolean>(false);

    // ---------------------PASSWORD----------------------
    const [passwordError, setPasswordError] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [passwordCheck, setPasswordCheck] = useState<string>('');
    const passwordInputRef = useRef<HTMLInputElement>(null);
    const passwordCheckInputRef = useRef<HTMLInputElement>(null);

    // ---------------------SIGNUP-------------------------
    const [signUpError, setSignUpError] = useState<string>('');

    const onChangeId = useCallback(
        handleInputChange(setId, setIdErrorMessage, validateId, () => {
            setSignUpError('');
            setIdSuccessMessage('');
            setIsIdDupliacted(false);
            setIsIdChecked(false);
        }),
        []
    );

    const onSubmitCheckId = useCallback(
        async (e: FormEvent<HTMLButtonElement>) => {
            e.preventDefault();
            setSignUpError('');
            // 입력 값 유효성 검사
            if (idErrorMessage || !id) {
                errorInputCheck(idInputRef.current);
                setIdErrorMessage(ValidationMessages.REQUIRED_ID);
                setIdSuccessMessage('');
                return;
            }

            try {
                const response = await checkId({ loginId: id });
                setIsIdChecked(true);
                setIdErrorMessage('');
                setIdSuccessMessage('');

                if (response) {
                    setIdSuccessMessage(ValidationMessages.CHECK_ID_SUCCESS);
                    setIsIdDupliacted(false);
                    return;
                }
            } catch (error) {
                if (error === 40002) {
                    setIdErrorMessage(ValidationMessages.EXIST_NICKNAME);
                    return;
                }
                setNicknameErrorMessage(
                    ValidationMessages.ERROR_CHECK_NICKNAME
                );
                return;
            }
        },
        [id, idErrorMessage]
    );

    const onChangeNickname = useCallback(
        handleInputChange(
            setNickname,
            setNicknameErrorMessage,
            validateNickname,
            () => {
                setSignUpError('');
                setNicknameSuccessMessage('');
                setIsNicknameChecked(false);
                setIsNicknameDupliacted(false);
            }
        ),
        []
    );

    const onSubmitCheckNickname = useCallback(
        async (e: FormEvent<HTMLButtonElement>) => {
            e.preventDefault();
            setSignUpError('');

            // 입력 값 유효성 검사
            if (nicknameErrorMessage || !nickname) {
                errorInputCheck(nicknameInputRef.current);
                setNicknameErrorMessage(ValidationMessages.REQUIRED_NICKNAME);
                setNicknameSuccessMessage('');
                return;
            }

            try {
                const response = await checkNickname({ nickname });
                setIsNicknameChecked(true);
                setNicknameErrorMessage('');
                setNicknameSuccessMessage('');

                // 닉네임 중복 처리
                if (response) {
                    setNicknameSuccessMessage(
                        ValidationMessages.CHECK_NICKNAME_SUCCESS
                    );
                    setIsNicknameDupliacted(false);
                    return;
                }
            } catch (error) {
                if (error === 40004) {
                    setNicknameErrorMessage(ValidationMessages.EXIST_NICKNAME);
                    return;
                }
                setNicknameErrorMessage(
                    ValidationMessages.ERROR_CHECK_NICKNAME
                );
                return;
            }
        },
        [nickname, nicknameErrorMessage]
    );

    const onChangePassword = useCallback(
        passwordCheckHandler(setPassword, passwordCheck, setPasswordError),
        [passwordCheck]
    );

    const onChangePasswordCheck = useCallback(
        passwordCheckHandler(setPasswordCheck, password, setPasswordError),
        [password]
    );

    const onClickVerification = useCallback(
        async (e: FormEvent<HTMLButtonElement>) => {
            e.preventDefault();
            if (!id) {
                errorInputCheck(idInputRef.current);
                return;
            }
            if (!isIdChecked) {
                setSignUpError(ValidationMessages.REQUIRED_CHECK_ID);
                return;
            }

            if (!nickname) {
                errorInputCheck(nicknameInputRef.current);
                return;
            }
            if (!isNicknameChecked) {
                setSignUpError(ValidationMessages.REQUIRED_CHECK_NICKNAME);
                return;
            }
            if (idErrorMessage) {
                errorInputCheck(idInputRef.current);
                return;
            }
            if (nicknameErrorMessage) {
                errorInputCheck(nicknameInputRef.current);
                return;
            }
            if (passwordError || !password) {
                errorInputCheck(passwordInputRef.current);
                return;
            }

            if (nickname && id && password && passwordCheck) {
                try {
                    const userData = {
                        loginId: id,
                        nickname: nickname,
                        password: password,
                    };
                    const response = await signUp(userData);
                    // 회원가입 후 자동 로그인
                    const loginData: LogInRequestDTO = {
                        loginId: id,
                        password: password,
                    };
                    await logIn(loginData);

                    auth.login(() => {
                        navigate('/signup/verification', { state: loginData });
                    });
                } catch (error: unknown) {
                    if (error === 40002) {
                        setSignUpError(ValidationMessages.EXIST_USER);
                        return;
                    }
                    setSignUpError(ValidationMessages.UNKNOWN_ERROR);
                }
            }
        },
        [
            id,
            password,
            nickname,
            idErrorMessage,
            nicknameErrorMessage,
            passwordError,
            signUpError,
            navigate,
        ]
    );

    const onSignUpSubmit = useCallback(
        async (e: FormEvent<HTMLFormElement>) => {
            e.preventDefault();
            if (!id) {
                errorInputCheck(idInputRef.current);
                return;
            }
            if (!isIdChecked) {
                setSignUpError(ValidationMessages.REQUIRED_CHECK_ID);
                return;
            }

            if (!nickname) {
                errorInputCheck(nicknameInputRef.current);
                return;
            }
            if (!isNicknameChecked) {
                setSignUpError(ValidationMessages.REQUIRED_CHECK_NICKNAME);
                return;
            }
            if (idErrorMessage) {
                errorInputCheck(idInputRef.current);
                return;
            }
            if (nicknameErrorMessage) {
                errorInputCheck(nicknameInputRef.current);
                return;
            }
            if (passwordError || !password) {
                errorInputCheck(passwordInputRef.current);
                return;
            }
            if (nickname && id && password && passwordCheck) {
                const userData: SignUpDTO = {
                    loginId: id,
                    nickname: nickname,
                    password: password,
                };
                try {
                    await signUp(userData);
                    // 회원가입 후 자동 로그인
                    const loginData: LogInRequestDTO = {
                        loginId: id,
                        password: password,
                    };
                    await logIn(loginData);

                    setSessionStorages({
                        key: StorageKeyword.CREATE_ACCOUNT_SUCCESS,
                        value: StorageKeyword.TRUE,
                    });

                    // 로그인 상태를 업데이트하고 홈 페이지로 리다이렉트
                    auth.login(() => {
                        navigate('/');
                    });
                } catch (error: unknown) {
                    if (error === 40002) {
                        setSignUpError(ValidationMessages.EXIST_USER);
                        return;
                    }
                    setSignUpError(ValidationMessages.UNKNOWN_ERROR);
                }
            }
        },
        [
            id,
            nickname,
            password,
            passwordCheck,
            idErrorMessage,
            passwordError,
            nicknameErrorMessage,
            signUpError,
        ]
    );

    return (
        <div className="main__container">
            <form className="c-login" onSubmit={onSignUpSubmit}>
                <div className="c-login__section">
                    <ErrorMessage message={idErrorMessage} />
                    <SuccessMessage message={idSuccessMessage} />
                    {!idErrorMessage && !idSuccessMessage && (
                        <p>{DEFAULT_ID}</p>
                    )}
                    <section className="c-login__section-verification">
                        <InputField
                            label="아이디"
                            type="text"
                            name="id"
                            value={id}
                            onChange={onChangeId}
                            placeholder={ValidationMessages.REQUIRED_ID}
                            ref={nicknameInputRef}
                            isDuplicated={isIdDupliacted}
                            isChecked={isIdChecked}
                            className={`c-login__input ${
                                isIdChecked
                                    ? isIdDupliacted
                                        ? 'fail'
                                        : 'success'
                                    : ''
                            }`}
                        />
                        <Button
                            className="button__rounded button__light"
                            type="button"
                            onClick={onSubmitCheckId}
                        >
                            {t('CHECK_DUPLICATED')}
                        </Button>
                    </section>
                </div>
                <div className="c-login__section">
                    <ErrorMessage message={nicknameErrorMessage} />
                    <SuccessMessage message={nicknameSuccessMessage} />
                    {!nicknameErrorMessage && !nicknameSuccessMessage && (
                        <p>{DEFAULT_NICKNAME}</p>
                    )}
                    <section className="c-login__section-verification">
                        <InputField
                            label="닉네임"
                            type="text"
                            name="nickname"
                            value={nickname}
                            onChange={onChangeNickname}
                            placeholder={t('REQUIRED_NICKNAME')}
                            ref={nicknameInputRef}
                            isDuplicated={isNicknameDupliacted}
                            isChecked={isNicknameChecked}
                            className={`c-login__input ${
                                isNicknameChecked
                                    ? isNicknameDupliacted
                                        ? 'fail'
                                        : 'success'
                                    : ''
                            }`}
                        />
                        {isNicknameChecked &&
                            (isNicknameDupliacted ? (
                                <i className="c-icon c-icon--fill-fail">
                                    close
                                </i>
                            ) : (
                                <i className="c-icon c-icon--fill-success">
                                    check
                                </i>
                            ))}
                        <Button
                            className="button__rounded button__light"
                            type="button"
                            onClick={onSubmitCheckNickname}
                        >
                            {t('CHECK_DUPLICATED')}
                        </Button>
                    </section>
                </div>
                <div className="c-login__section">
                    <ErrorMessage message={passwordError} />
                    {!passwordError && <p>{DEFAULT_PASSWORD}</p>}
                    <section className="c-login__section-password">
                        <InputField
                            label="비밀번호"
                            type="password"
                            name="password"
                            value={password}
                            onChange={onChangePassword}
                            placeholder={ValidationMessages.DEFAULT_PASSWORD}
                            ref={passwordInputRef}
                            className="c-login__input"
                        />
                        <InputField
                            label="비밀번호 확인"
                            type="password"
                            name="password-check"
                            value={passwordCheck}
                            onChange={onChangePasswordCheck}
                            placeholder={
                                ValidationMessages.DEFAULT_PASSWORD_CHECK
                            }
                            ref={passwordCheckInputRef}
                            className="c-login__input"
                        />
                    </section>
                </div>
                {signUpError && (
                    <p className="message font-warning">{signUpError}</p>
                )}
                <section className="c-login__button-section">
                    <div className="c-login__button-section-message">
                        <p>{t('VERIFICATION_NOTICE-1')}</p>
                        <p>{t('VERIFICATION_NOTICE-2')}</p>
                        <p>{t('VERIFICATION_NOTICE-3')}</p>
                    </div>
                    <Button
                        className="button__rounded button__orange"
                        type="button"
                        onClick={onClickVerification}
                    >
                        {t('VERIFICATION_NAVIGATE_BUTTON')}
                    </Button>
                    <Button
                        className="button__rounded button__light"
                        type="submit"
                    >
                        {t('DEFAULT_SIGNUP')}
                    </Button>
                </section>
            </form>
        </div>
    );
}
