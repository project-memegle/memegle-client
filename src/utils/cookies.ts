// cookies.ts

// 쿠키 저장 함수
export const setCookie = (name: string, value: string, days?: number): void => {
    let expires = '';
    if (days) {
        const date = new Date();
        date.setTime(date.getTime() + days * 24 * 60 * 60 * 1000);
        expires = `; expires=${date.toUTCString()}`;
    }
    document.cookie = `${name}=${
        value || ''
    }${expires}; path=/; secure; samesite=strict`;
};

// 쿠키 가져오기 함수
export const getCookie = (name: string): string | null => {
    const nameEQ = `${name}=`;
    const cookiesArray = document.cookie.split(';');
    for (let i = 0; i < cookiesArray.length; i++) {
        let cookie = cookiesArray[i];
        while (cookie.charAt(0) === ' ') {
            cookie = cookie.substring(1, cookie.length);
        }
        if (cookie.indexOf(nameEQ) === 0) {
            return cookie.substring(nameEQ.length, cookie.length);
        }
    }
    return null;
};

// 쿠키 삭제 함수
export const deleteCookie = (name: string): void => {
    document.cookie = `${name}=; Max-Age=-99999999; path=/`;
};
