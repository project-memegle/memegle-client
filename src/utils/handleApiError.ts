import axios, { AxiosError } from 'axios';
import ValidationMessages from '../components/Validations/ValidationMessages';

const errorMessages: Record<number, string> = {
    40000: ValidationMessages.LOGIN_FAILED,
    40001: ValidationMessages.INVALID_FORM,
    40002: ValidationMessages.EXIST_USER,
    40003: ValidationMessages.MISSED_FORM,
    40102: ValidationMessages.INVALID_PASSWORD_LENGTH,
    40401: ValidationMessages.MISSED_RESOURCE,
    50000: ValidationMessages.SERVER_ERROR,
};

export const handleAxiosError = (
    error: AxiosError,
    setMessage: (message: string) => void
) => {
    if (axios.isAxiosError(error)) {
        const status = error.response?.status;

        const message =
            status !== undefined
                ? errorMessages[status]
                : ValidationMessages.UNKNOWN_ERROR;
        setMessage(message);
    } else {
        setMessage(ValidationMessages.UNKNOWN_ERROR);
    }
};

const networkErrorMessages: Record<string, string> = {
    ECONNABORTED: ValidationMessages.TIMEOUT_ERROR,
    ENETUNREACH: ValidationMessages.NETWORK_ERROR,
    ECONNREFUSED: ValidationMessages.CONNECTION_REFUSED,
};

const handleNetworkError = (
    error: Error,
    setMessage: (message: string) => void
) => {
    const message = networkErrorMessages[error.message];
    setMessage(message);
};

export const handleApiError = (
    error: AxiosError | Error | unknown,
    setMessage: (message: string) => void
) => {
    if (axios.isAxiosError(error)) {
        handleAxiosError(error, setMessage);
    } else if (error instanceof Error) {
        handleNetworkError(error, setMessage);
    } else {
        setMessage(ValidationMessages.UNKNOWN_ERROR);
    }
};
