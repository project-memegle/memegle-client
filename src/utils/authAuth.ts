import { getCookie } from './cookies';

const ACCESS_TOKEN = 'access_token';

export const getAccessToken = (): string | null => {
    return getCookie(ACCESS_TOKEN);
};
