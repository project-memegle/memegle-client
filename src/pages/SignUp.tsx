import { FormEvent, useCallback, useRef, useState } from 'react';
import { AxiosError } from 'axios';
import validateId from 'components/Validations/ValidateId';
import ValidationMessages from 'components/Validations/ValidationMessages';
import validateNickname from 'components/Validations/ValidateNickname';
import { SignUpDTO } from 'services/dto/SignUpDto';
import { errorInputCheck } from 'utils/Event/errorInputCheck';
import handleInputChange from 'utils/Event/handleInputChange';
import passwordCheckHandler from 'utils/SignUp/passwordCheckHandler';
import { post } from 'utils/API/fetcher';
import { handleApiError } from 'utils/API/handleApiError';
import { setSessionStorages } from 'utils/Storage/sessionStorage';
import ToastMessage from 'components/UI/ToastMessage/ToastMessage';
import useCustomNavigate from 'hooks/useCustomNaviaget';
import { signUp } from 'services/SignupService';
import { LogInRequestDTO } from 'services/dto/LogInDto';
import { logIn } from 'services/LogInService';
import { useAuth } from 'components/auth/ProvideAuth';

export default function SignUp() {
    const navigate = useCustomNavigate();
    const auth = useAuth();

    const DEFAULT_ID = ValidationMessages.DEFAULT_ID;
    const DEFALUT_NICKNAME = ValidationMessages.DEFAULT_NICKNAME;
    const DEFAULT_PASSWORD = ValidationMessages.DEFAULT_PASSWORD;

    const [idError, setIdError] = useState(DEFAULT_ID);
    const [nicknameError, setNicknameError] = useState(DEFALUT_NICKNAME);
    const [passwordError, setPasswordError] = useState(DEFAULT_PASSWORD);

    const [toastMessage, setToastMessage] = useState('');
    const [toast, setToast] = useState(false);

    const [id, setId] = useState('');
    const [nickname, setNickname] = useState('');
    const [password, setPassword] = useState('');
    const [passwordCheck, setPasswordCheck] = useState('');
    const [signUpError, setSignUpError] = useState('');
    const [signupSuccess, setSignupSuccess] = useState('');

    const idInputRef = useRef<HTMLInputElement>(null);
    const nicknameInputRef = useRef<HTMLInputElement>(null);
    const passwordInputRef = useRef<HTMLInputElement>(null);
    const passwordCheckInputRef = useRef<HTMLInputElement>(null);

    const onChangeId = useCallback(
        handleInputChange(setId, setIdError, validateId, () =>
            setSignupSuccess('')
        ),
        []
    );

    const onChangeNickname = useCallback(
        handleInputChange(setNickname, setNicknameError, validateNickname, () =>
            setSignupSuccess('')
        ),
        []
    );

    const onChangePassword = useCallback(
        passwordCheckHandler(setPassword, passwordCheck, setPasswordError, () =>
            setSignupSuccess('')
        ),
        [passwordCheck]
    );

    const onChangePasswordCheck = useCallback(
        passwordCheckHandler(setPasswordCheck, password, setPasswordError, () =>
            setSignupSuccess('')
        ),
        [password]
    );

    const onClickVerification = useCallback(
        async (e: FormEvent<HTMLButtonElement>) => {
            e.preventDefault();
            if (idError || nicknameError || passwordError) {
                if (idError) errorInputCheck(idInputRef.current);
                else if (nicknameError)
                    errorInputCheck(nicknameInputRef.current);
                else if (passwordError)
                    errorInputCheck(passwordInputRef.current);
                signUpError && setSignUpError(ValidationMessages.SIGNUP_ERROR);
                return;
            }

            if (nickname && id && password && passwordCheck) {
                try {
                    setSessionStorages({ key: 'id', value: id });
                    setSessionStorages({ key: 'password', value: password });
                    setSessionStorages({ key: 'nickname', value: nickname });
                    navigate('/verification');
                } catch (error) {
                    setToastMessage(ValidationMessages.FAILED_EVENT);
                    setToast(true);
                }
            }
        },
        [
            id,
            password,
            nickname,
            idError,
            nicknameError,
            passwordError,
            signUpError,
            navigate,
        ]
    );

    const onSubmit = useCallback(
        async (e: FormEvent<HTMLFormElement>) => {
            e.preventDefault();
            if (idError || nicknameError || passwordError) {
                if (idError) errorInputCheck(idInputRef.current);
                else if (nicknameError)
                    errorInputCheck(nicknameInputRef.current);
                else if (passwordError)
                    errorInputCheck(passwordInputRef.current);
                signUpError && setSignUpError(ValidationMessages.SIGNUP_ERROR);
                return;
            }

            if (nickname && id && password && passwordCheck) {
                const userData: SignUpDTO = {
                    loginId: id,
                    nickname: nickname,
                    password: password,
                };
                try {
                    const response = await signUp(userData);
                    console.log(response);
                    // 회원가입 후 자동 로그인
                    const loginData: LogInRequestDTO = {
                        loginId: id,
                        password: password,
                    };
                    await logIn(loginData);

                    // 로그인 상태를 업데이트하고 홈 페이지로 리다이렉트
                    auth.login(() => {
                        navigate('/');
                    });
                } catch (error) {
                    handleApiError(error as AxiosError, setSignUpError);
                }
            }
        },
        [
            id,
            nickname,
            password,
            passwordCheck,
            idError,
            passwordError,
            nicknameError,
            signUpError,
        ]
    );

    return (
        <div className="main__container">
            <form className="c-login" onSubmit={onSubmit}>
                <div className="c-login__section">
                    <p>{idError ? idError : DEFAULT_ID}</p>
                    <label htmlFor="id">아이디</label>
                    <input
                        ref={idInputRef}
                        className="c-login__input"
                        name="id"
                        id="id"
                        type="text"
                        placeholder="아이디"
                        value={id}
                        onChange={onChangeId}
                    />
                </div>
                <div className="c-login__section">
                    <p>{nicknameError ? nicknameError : DEFALUT_NICKNAME}</p>
                    <label htmlFor="nickname">닉네임</label>
                    <input
                        ref={nicknameInputRef}
                        className="c-login__input"
                        name="nickname"
                        id="nickname"
                        type="text"
                        placeholder="닉네임"
                        value={nickname}
                        onChange={onChangeNickname}
                    />
                </div>
                <div className="c-login__section">
                    <p>{passwordError ? passwordError : DEFAULT_PASSWORD}</p>
                    <section className="c-login__section-password">
                        <div>
                            <label htmlFor="password">비밀번호</label>
                            <input
                                ref={passwordInputRef}
                                className="c-login__input"
                                name="password"
                                type="password"
                                id="password"
                                placeholder="비밀번호"
                                value={password}
                                onChange={onChangePassword}
                            />
                        </div>
                        <div>
                            <label htmlFor="password-check">
                                비밀번호 확인
                            </label>
                            <input
                                ref={passwordCheckInputRef}
                                className="c-login__input"
                                name="password-check"
                                type="password"
                                id="password-check"
                                placeholder="비밀번호 확인"
                                value={passwordCheck}
                                onChange={onChangePasswordCheck}
                            />
                        </div>
                    </section>
                </div>
                {signUpError && <p className="message">{signUpError}</p>}
                {signupSuccess && <p className="message">{signupSuccess}</p>}
                <section className="c-login__button-section">
                    <div>
                        <p>
                            본인 인증을 완료한 회원만이 아이디 및 비밀번호 찾기
                            서비스를 이용할 수 있습니다.
                        </p>
                        <p>본인 인증 없이도 회원 가입이 가능합니다.</p>
                    </div>
                    <button
                        className="button__rounded button__orange"
                        type="button"
                        onClick={onClickVerification}
                    >
                        본인인증
                    </button>
                    <button className="button__rounded button__light">
                        회원가입
                    </button>
                </section>
            </form>
            {toast && (
                <ToastMessage
                    message={toastMessage}
                    onClose={() => setToast(false)}
                />
            )}
        </div>
    );
}
