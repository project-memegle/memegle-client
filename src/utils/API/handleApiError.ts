import axios, { AxiosError } from 'axios';
import { handleErrorPage } from '../../pages/Error';
import getValidationMessages from '../../components/Validations/ValidationMessages';
const ValidationMessages = getValidationMessages();

const AxiosErrorMessages: Record<number, string> = {
    40000: ValidationMessages.REQUIRED_FORM,
    40001: ValidationMessages.INVALID_FORM,
    40002: ValidationMessages.EXIST_ID,
    40003: ValidationMessages.MISSING_FORM,
    40100: ValidationMessages.UNMATCHED_TOKEN,
    40101: ValidationMessages.INVALID_TOKEN,
    40102: ValidationMessages.INVALID_PASSWORD_LENGTH,
    40401: ValidationMessages.MISSING_RESOURCE,
    50000: ValidationMessages.SERVER_ERROR,
};

export const handleApiError = (
    error: AxiosError | unknown,
    setMessage?: (message: string) => void
) => {
    let message: string;

    if (!axios.isAxiosError(error) || !error.response) {
        message = ValidationMessages.UNKNOWN_ERROR;
        handleErrorPage(message);
        if (setMessage) {
            setMessage(message);
        }
        return;
    }

    const status = error.response.status;
    const statusCode = error.response.data.code;

    if (status === 200 || status === 204) {
        return;
    }

    message =
        AxiosErrorMessages[statusCode] || ValidationMessages.UNKNOWN_ERROR;

    handleErrorPage(message);
    if (setMessage) {
        setMessage(message);
    }
    return message;
};
