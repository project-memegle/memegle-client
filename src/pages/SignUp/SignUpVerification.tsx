import { FormEvent, useCallback, useEffect, useState } from 'react';
import { AxiosError } from 'axios';
import { handleApiError } from 'utils/API/handleApiError';
import useCustomNavigate from 'hooks/useCustomNaviaget';
import { setSessionStorages } from 'utils/Storage/sessionStorage';
import { verifyVerificationCode } from 'services/VerificationService';
import StorageKeyword from 'Constant/StorageKeyword';
import { useLocation } from 'react-router-dom';
import { useAuth } from 'components/auth/ProvideAuth';
import { LogInRequestDTO } from 'services/dto/LogInDto';
import { logIn } from 'services/LogInService';
import { VerifyCodePasswordDTO } from 'services/dto/PasswordDto';
import VerificationForm from 'components/UI/VerificationForm';

export default function SignUpVerification() {
    const navigate = useCustomNavigate();
    const auth = useAuth();

    const location = useLocation();
    const logInUserData = location.state as LogInRequestDTO;
    const [message, setMessage] = useState('');

    useEffect(() => {
        if (!logInUserData) {
            navigate('/signup');
        }
    }, [logInUserData, navigate]);

    const skipVerification = async () => {
        if (!logInUserData) {
            navigate('/signup');
        }
        try {
            const loginData: LogInRequestDTO = {
                loginId: logInUserData.loginId,
                password: logInUserData.password,
            };

            await logIn(loginData);

            setSessionStorages({
                key: StorageKeyword.CREATE_ACCOUNT_SUCCESS,
                value: StorageKeyword.TRUE,
            });

            auth.login(() => {
                navigate('/');
            });
        } catch (error) {
            handleApiError(error as AxiosError, setMessage);
        }
    };

    const handleSubmit = useCallback(
        async (name: string, email: string, code: string) => {
            if (logInUserData) {
                const userData: VerifyCodePasswordDTO = {
                    email: email,
                    authenticationCode: code,
                    authenticationType: StorageKeyword.VERIFICATION_CODE_SIGNUP,
                };
                setMessage('');
                try {
                    await verifyVerificationCode(userData);
                    skipVerification();

                    const { loginId, password } = logInUserData;

                    const loginData: LogInRequestDTO = {
                        loginId,
                        password,
                    };
                    await logIn(loginData);

                    setSessionStorages({
                        key: StorageKeyword.CREATE_ACCOUNT_SUCCESS,
                        value: StorageKeyword.TRUE,
                    });

                    auth.login(() => {
                        navigate('/');
                    });
                } catch (error) {
                    handleApiError(error as AxiosError, setMessage);
                }
            }
        },
        [logInUserData, navigate, auth]
    );

    return (
        <div className="main__container">
            <VerificationForm
                onSubmit={handleSubmit}
                onSkipVerification={skipVerification}
            />
            {message && <p className="message">{message}</p>}
        </div>
    );
}
