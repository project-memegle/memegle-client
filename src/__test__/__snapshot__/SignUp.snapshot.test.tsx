import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import axios from 'axios';
import SignUp from '../../pages/SignUp';
import { handleApiError } from '../../utils/handleApiError';

jest.mock('../util/handleApiError', () => ({
    handleApiError: jest.fn(),
}));

jest.mock('axios'); // Mock axios

describe('SignUp Component', () => {
    beforeEach(() => {
        jest.clearAllMocks(); // Clear previous mocks before each test
    });

    test('renders SignUp form', () => {
        render(<SignUp />);

        // Check if input fields and button are in the document
        expect(screen.getByLabelText(/아이디/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/닉네임/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/비밀번호/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/비밀번호 확인/i)).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /회원가입/i })
        ).toBeInTheDocument();
    });

    test('submits form with valid data', async () => {
        render(<SignUp />);

        // Fill in the form fields
        fireEvent.change(screen.getByLabelText(/아이디/i), {
            target: { value: 'testUser' },
        });
        fireEvent.change(screen.getByLabelText(/닉네임/i), {
            target: { value: 'testNickname' },
        });
        fireEvent.change(screen.getByLabelText(/비밀번호/i), {
            target: { value: 'TestPassword123!' },
        });
        fireEvent.change(screen.getByLabelText(/비밀번호 확인/i), {
            target: { value: 'TestPassword123!' },
        });

        // Mock the Axios post response
        (axios.post as jest.Mock).mockResolvedValueOnce({ data: {} });

        // Submit the form
        fireEvent.click(screen.getByRole('button', { name: /회원가입/i }));

        // Wait for the success message
        await waitFor(() => {
            expect(screen.getByText(/회원가입 성공/i)).toBeInTheDocument(); // Adjust according to your success message
        });

        expect(axios.post).toHaveBeenCalledWith('/signup', {
            loginId: 'testUser',
            nickname: 'testNickname',
            password: 'TestPassword123!',
        });
    });

    test('shows error message on API failure', async () => {
        render(<SignUp />);

        // Fill in the form fields
        fireEvent.change(screen.getByLabelText(/아이디/i), {
            target: { value: 'testUser' },
        });
        fireEvent.change(screen.getByLabelText(/닉네임/i), {
            target: { value: 'testNickname' },
        });
        fireEvent.change(screen.getByLabelText(/비밀번호/i), {
            target: { value: 'TestPassword123!' },
        });
        fireEvent.change(screen.getByLabelText(/비밀번호 확인/i), {
            target: { value: 'TestPassword123!' },
        });

        // Mock the Axios post to reject with an error
        const errorMessage = 'Failed to sign up';
        (axios.post as jest.Mock).mockRejectedValueOnce(
            new Error(errorMessage)
        );

        // Submit the form
        fireEvent.click(screen.getByRole('button', { name: /회원가입/i }));

        // Wait for the error message
        await waitFor(() => {
            expect(handleApiError).toHaveBeenCalled(); // Check if error handling was called
            expect(screen.getByText(errorMessage)).toBeInTheDocument(); // Adjust according to your error message handling
        });
    });
});
