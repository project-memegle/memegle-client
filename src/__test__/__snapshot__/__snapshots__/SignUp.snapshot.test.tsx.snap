// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`회원가입 컴포넌트 렌더링 테스트 1`] = `
<DocumentFragment>
  <div
    class="main__container"
  >
    <form
      class="c-login"
    >
      <div
        class="c-login__section"
      >
        <p>
          아이디
        </p>
        <label
          for="id"
        >
          아이디
        </label>
        <input
          class="c-login__input"
          id="id"
          name="id"
          placeholder="아이디"
          type="text"
          value=""
        />
      </div>
      <div
        class="c-login__section"
      >
        <p>
          닉네임
        </p>
        <label
          for="nickname"
        >
          닉네임
        </label>
        <input
          class="c-login__input"
          id="nickname"
          name="nickname"
          placeholder="닉네임"
          type="text"
          value=""
        />
      </div>
      <div
        class="c-login__section"
      >
        <p>
          비밀번호
        </p>
        <section
          class="c-login__section-password"
        >
          <div>
            <label
              for="password"
            >
              비밀번호
            </label>
            <input
              class="c-login__input"
              id="password"
              name="password"
              placeholder="비밀번호"
              type="password"
              value=""
            />
          </div>
          <div>
            <label
              for="password-check"
            >
              비밀번호 확인
            </label>
            <input
              class="c-login__input"
              id="password-check"
              name="password-check"
              placeholder="비밀번호 확인"
              type="password"
              value=""
            />
          </div>
        </section>
      </div>
      <section
        class="c-login__button-section"
      >
        <button
          class="button__rounded button__light"
          type="submit"
        >
          회원가입
        </button>
      </section>
    </form>
  </div>
</DocumentFragment>
`;
